### PR TITLE
Add Experimental Initial Span

### PR DESCRIPTION
**Pull request by AhmedZero**

### CODE REVIEW

File: src/Generator/Generators/CSharp/CSharpTypePrinter.cs [hunk 1]
```diff
@@ -122,7 +122,11 @@ public override TypePrinterResult VisitTagType(TagType tag, TypeQualifiers quals
                     "[MarshalAs(UnmanagedType.LPArray)] ";
                 return $"{prefix}string[]";
             }
+            if (!(array.SizeType != ArrayType.ArraySize.Constant &&  MarshalKind == MarshalKind.ReturnVariableArray) && Context.Options.UseSpan)
```
Comment (tritao) [inline src/Generator/Generators/CSharp/CSharpTypePrinter.cs:125]: This is missing a new line here before the if.

File: src/Generator/Generators/CSharp/CSharpTypePrinter.cs [hunk 2]
```diff
@@ -122,7 +122,11 @@ public override TypePrinterResult VisitTagType(TagType tag, TypeQualifiers quals
...
+            {
+                return $"Span<{arrayType.Visit(this)}>";
```
Comment (tritao) [inline src/Generator/Generators/CSharp/CSharpTypePrinter.cs:128]: Remove the extra line here.

File: src/Generator/Options.cs [hunk 3]
```diff
@@ -184,6 +184,9 @@ public Module AddModule(string libraryName)
         public readonly List<string> DependentNameSpaces = new List<string>();
         public bool MarshalCharAsManagedChar { get; set; }
 
+        public bool UseSpan { get; set; }
```
Comment (tritao) [inline src/Generator/Options.cs:187]: Please document this option with XML comments and remove the extra new lines.

File: src/Generator/Generators/CSharp/CSharpMarshal.cs [hunk 4]
```diff
@@ -69,11 +69,18 @@ public override bool VisitArrayType(ArrayType array, TypeQualifiers quals)
                     var arrayType = array.Type.Desugar();
 
                     if (CheckIfArrayCanBeCopiedUsingMemoryCopy(array))
-                        Context.Return.Write($"CppSharp.Runtime.MarshalUtil.GetArray<{arrayType}>({Context.ReturnVarName}, {array.Size})");
+                        if (Context.Context.Options.UseSpan)
```
Comment (tritao) [inline src/Generator/Generators/CSharp/CSharpMarshal.cs:72]: Please add brackets to the if here. The way its nested can lead to bugs.

File: src/Generator/Generators/CSharp/CSharpMarshal.cs [hunk 5]
```diff
@@ -69,11 +69,18 @@ public override bool VisitArrayType(ArrayType array, TypeQualifiers quals)
...
+                            Context.Return.Write($"new Span<{arrayType}>({Context.ReturnVarName}, {array.Size})");
+
+                        else
+                            Context.Return.Write($"CppSharp.Runtime.MarshalUtil.GetArray<{arrayType}>({Context.ReturnVarName}, {array.Size})");
                     else if (array.Type.IsPrimitiveType(PrimitiveType.Char) && Context.Context.Options.MarshalCharAsManagedChar)
                         Context.Return.Write($"CppSharp.Runtime.MarshalUtil.GetCharArray({Context.ReturnVarName}, {array.Size})");
                     else if (array.Type.IsPointerToPrimitiveType(PrimitiveType.Void))
```
Comment (tritao) [inline src/Generator/Generators/CSharp/CSharpMarshal.cs:79]: Please add brackets to the if here. The way its nested can lead to bugs.

**Review by tritao — CHANGES_REQUESTED** [review]

**Comment by AhmedZero** [comment]
Review again thanks.

**Comment by tritao** [comment]
Thanks!

**Pull request merged**